File: loopx/control_plane/agents/capability_memory.ts
/** Host-local observations, never Goal configuration or an authorization grant. */
import {createHash} from "node:crypto";
import {readFile} from "node:fs/promises";
import {join} from "node:path";
import type {JsonObject} from "../effect_program.ts";
import {durableWriteJson, withFileMutationLock} from "../effect_runtime_io.ts";
import {requireBoolean, requireJsonObject, requireNonEmptyString, requireStringArray} from "../runtime_decode.ts";
import {OBSERVABLE_RUNTIME_CAPABILITIES} from "./capability_gate.ts";

const SCHEMA = "agent_runtime_capabilities_v0";
const observable = (value: string) => OBSERVABLE_RUNTIME_CAPABILITIES.has(value);
const lexical = (left: string, right: string) => left.localeCompare(right);
const unique = (values: string[]) => [...new Set(values)].sort(lexical);
const sortedEntries = (value: JsonObject) =>
  Object.entries(value).sort(([left], [right]) => lexical(left, right));

export async function agentCapabilityMemory(request: JsonObject): Promise<JsonObject> {
  if (request.schema_version !== "agent_runtime_capability_request_v0") {
    throw new TypeError("agent runtime capability request schema mismatch");
  }
  const root = requireNonEmptyString(request.runtime_root, "runtime_root");
  const scope = {
    registry: requireNonEmptyString(request.registry, "registry"),
    goal_id: requireNonEmptyString(request.goal_id, "goal_id"),
    agent_id: requireNonEmptyString(request.agent_id, "agent_id"),
  };
  const registered = requireStringArray(request.registered_agents, "registered_agents");
  if (!registered.includes(scope.agent_id)) throw new TypeError("agent must be registered for this Goal");
  const supplied = unique(requireStringArray(request.available ?? [], "available"));
  const available = supplied.filter(observable);
  const unavailable = unique(requireStringArray(request.unavailable ?? [], "unavailable"));
  const forget = unique(requireStringArray(request.forget ?? [], "forget"));
  if ([...unavailable, ...forget].some(c => !observable(c))) {
    throw new TypeError("only observable runtime capabilities can be marked unavailable or forgotten");
  }
  if (available.some(c => unavailable.includes(c) || forget.includes(c)) || unavailable.some(c => forget.includes(c))) {
    throw new TypeError("capability updates must not contradict each other");
  }
  const execute = requireBoolean(request.execute, "execute");
  const key = createHash("sha256").update(JSON.stringify(scope)).digest("hex");
  const path = join(root, "agent-runtime-capabilities", `${key}.json`);
  const read = async (): Promise<JsonObject> => {
    let bytes: string;
    try { bytes = await readFile(path, "utf8"); }
    catch (error) {
      if ((error as NodeJS.ErrnoException).code === "ENOENT") return {};
      throw error;
    }
    const state = requireJsonObject(JSON.parse(bytes), "agent runtime capability state");
    if (state.schema_version !== SCHEMA || Object.entries(scope).some(([k, v]) => state[k] !== v)) {
      throw new TypeError("agent runtime capability state scope/schema mismatch");
    }
    const observations = requireJsonObject(state.observations, "observations");
    for (const [capability, value] of Object.entries(observations)) {
      if (!observable(capability) || (value !== "available" && value !== "unavailable")) {
        throw new TypeError("invalid agent runtime capability observation");
      }
    }
    return observations;
  };
  const update = async (): Promise<JsonObject> => {
    const previous = await read();
    const next = {...previous};
    for (const c of available) next[c] = "available";
    for (const c of unavailable) next[c] = "unavailable";
    for (const c of forget) delete next[c];
    const changed = JSON.stringify(sortedEntries(previous)) !== JSON.stringify(sortedEntries(next));
    if (execute && changed) await durableWriteJson(path, {schema_version: SCHEMA, ...scope, observations: next});
    const project = (observations: JsonObject) => ({
      available: Object.keys(observations).filter(c => observations[c] === "available").sort(lexical),
      unavailable: Object.keys(observations).filter(c => observations[c] === "unavailable").sort(lexical),
    });
    return {
      schema_version: SCHEMA, goal_id: scope.goal_id, agent_id: scope.agent_id,
      scope: "host_registry_goal_agent", source: "agent_runtime_observation",
      ...project(execute ? next : previous),
      ...(changed && !execute ? {proposed: project(next)} : {}),
      invocation_only: supplied.filter(c => !observable(c)),
      written: execute && changed, durable_grant_written: false,
    };
  };
  return execute && (available.length || unavailable.length || forget.length)
    ? withFileMutationLock(path, update) : update();
}

File: loopx/control_plane/agents/capability_memory.py
"""Adapt registered host-local Agent observations to the typed state owner."""
from pathlib import Path
from typing import Any

from ...agent_registry import load_goal_from_registry, registered_agent_ids_for_goal
from ..effect_runtime import effect_runtime_result
from ..todos.contract import normalize_required_capabilities, normalize_todo_claimed_by


def agent_capability_memory(
    *, registry_path: Path, runtime_root: Path, goal_id: str, agent_id: str,
    available: Any = None, unavailable: Any = None, forget: Any = None,
    execute: bool = False,
) -> dict[str, Any]:
    goal_id = str(goal_id).strip()
    normalized_agent = normalize_todo_claimed_by(agent_id)
    if not normalized_agent:
        raise ValueError("agent_id must be a public-safe registered agent id")
    goal = load_goal_from_registry(registry_path, goal_id)
    result = effect_runtime_result("agent.capability_memory", {
        "schema_version": "agent_runtime_capability_request_v0",
        "runtime_root": str(runtime_root.expanduser().resolve()),
        "registry": str(registry_path.expanduser().resolve()),
        "goal_id": goal_id, "agent_id": normalized_agent,
        "registered_agents": registered_agent_ids_for_goal(goal),
        "available": normalize_required_capabilities(available),
        "unavailable": normalize_required_capabilities(unavailable),
        "forget": normalize_required_capabilities(forget),
        "execute": execute,
    })
    if not isinstance(result, dict) or result.get("schema_version") != "agent_runtime_capabilities_v0":
        raise TypeError("invalid agent runtime capability memory result")
    return result


def turn_start_hook(
    *,
    registry_path: Path,
    runtime_root: Path,
    goal_id: str,
    agent_id: str,
    available: Any,
):
    """Persist invocation observations through the admitted turn-start hook."""

    from ..capability_hooks import (
        TURN_START_HOOK_RESULT_SCHEMA_VERSION,
        TurnStartHookRegistration,
    )

    normalized = normalize_required_capabilities(available)

    def produce() -> dict[str, Any]:
        state = agent_capability_memory(
            registry_path=registry_path,
            runtime_root=runtime_root,
            goal_id=goal_id,
            agent_id=agent_id,
            available=normalized,
            execute=True,
        )
        return {
            "schema_version": TURN_START_HOOK_RESULT_SCHEMA_VERSION,
            "hook_id": "agent.capability_memory",
            "capability_id": "agent-capability-memory",
            "phase": "turn_start",
            "status": "empty",
            "observation_count": 0,
            "agent_read_required": False,
            "external_reads_performed": False,
            "external_writes_performed": False,
            "local_private_state_mutated": state.get("written") is True,
            "private_content_returned": False,
            "provider_payload_returned": False,
            "error_code": None,
        }

    return TurnStartHookRegistration(
        hook_id="agent.capability_memory",
        capability_id="agent-capability-memory",
        requested_read_scope=("goal_agent_registration",),
        requested_write_scope=("agent_private_capability_memory",),
        producer=produce,
    )


def extend_turn_start_dispatch(
    dispatch: dict[str, Any],
    *,
    registry_path: Path,
    runtime_root: Path,
    goal_id: str,
    agent_id: str | None,
    available: Any,
) -> dict[str, Any]:
    """Compose capability memory without putting provider writes in quota core."""

    normalized = normalize_required_capabilities(available)
    if not agent_id or not normalized:
        return dispatch
    from ..capability_hooks import dispatch_turn_start_hooks

    extra = dispatch_turn_start_hooks(
        (
            turn_start_hook(
                registry_path=registry_path,
                runtime_root=runtime_root,
                goal_id=goal_id,
                agent_id=agent_id,
                available=normalized,
            ),
        )
    )
    result = dict(dispatch)
    for key in ("results", "required_reads", "failures"):
        result[key] = list(dispatch.get(key) or []) + list(extra.get(key) or [])
    for key in ("registered_count", "invoked_count"):
        result[key] = int(dispatch.get(key) or 0) + int(extra.get(key) or 0)
    return result


def resolve_agent_capabilities(
    status_payload: dict[str, Any], *, goal_id: str, agent_identity: dict[str, Any] | None,
    item: dict[str, Any], project_asset: dict[str, Any], available: Any,
    state: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Read current scoped observations even when the caller's status is cached."""
    from .capability_gate import _evaluate
    from ..quota.goal_boundary import declared_available_capabilities

    root, registry = status_payload.get("runtime_root"), status_payload.get("registry")
    if state is None and agent_identity and root and registry:
        state = agent_capability_memory(
            registry_path=Path(str(registry)), runtime_root=Path(str(root)),
            goal_id=goal_id, agent_id=agent_identity["agent_id"],
        )
    availability = _evaluate(
        "availability",
        goal=[*declared_available_capabilities(item), *declared_available_capabilities(project_asset)],
        runtime=normalize_required_capabilities(available), agent=state or {},
    )
    if agent_identity:
        # Default status/quota surfaces need only the compact runtime input used
        # to rebuild interaction contracts. Detailed inheritance and correction
        # readback belongs to the explicit ``agent-capabilities`` command.
        agent_identity["runtime_available_capabilities"] = availability[
            "runtime_available"
        ]
    return availability
